ml/jml/onevsall_generator.cc

```diff
@@ -0,0 +1,139 @@
+// This file is part of MLDB. Copyright 2017 mldb.ai inc. All rights reserved.
+
+/* onevsall_generator.cc                                          -*- C++ -*-
+   Mathieu Marquis Bolduc, 8 March 2017
+   Copyright (c) 2017 MLDB.ai  All rights reserved.
+   $Source$
+
+   Generator for onevsall classifiers
+*/
+
+#include "onevsall_generator.h"
+#include "mldb/ml/jml/registry.h"
+#include "mldb/jml/utils/smart_ptr_utils.h"
+
+#include "training_index.h"
+
+using namespace std;
+
+
+namespace ML {
+
+/*****************************************************************************/
+/* ONEVSALL_GENERATOR                                                        */
+/*****************************************************************************/
+
+OneVsAll_Classifier_Generator::
+OneVsAll_Classifier_Generator()
+{
+    defaults();
+}
+
+OneVsAll_Classifier_Generator::
+OneVsAll_Classifier_Generator(std::shared_ptr<Classifier_Generator> learner)
+{
+    defaults();
+    weak_learner = learner;
+}
+
+OneVsAll_Classifier_Generator::~OneVsAll_Classifier_Generator()
+{
+}
+
+void
+OneVsAll_Classifier_Generator::
+configure(const Configuration & config, vector<string> & unparsedKeys)
+{
+    Classifier_Generator::configure(config, unparsedKeys);
+
+    weak_learner = get_trainer("weak_learner", config);
+}
+
+void
+OneVsAll_Classifier_Generator::
+defaults()
+{
+    Classifier_Generator::defaults();
+    weak_learner.reset();
+}
+
+Config_Options
+OneVsAll_Classifier_Generator::
+options() const
+{
+    Config_Options result = Classifier_Generator::options();
+    result       
+        .subconfig("weak_leaner", weak_learner,
+                   "Binary classifier for each label value");
+    
+    return result;
+}
+
+void
+OneVsAll_Classifier_Generator::
+init(std::shared_ptr<const Feature_Space> fs, Feature predicted)
+{
+    Classifier_Generator::init(fs, predicted);
+    model = OneVsAllClassifier(fs, predicted);    
+}
+
+std::shared_ptr<Classifier_Impl>
+OneVsAll_Classifier_Generator::
+generate(Thread_Context & context,
+         const Training_Data & training_data,
+         const distribution<float> & weights,
+         const std::vector<Feature> & features,
+         int) const
+{
+    std::shared_ptr<OneVsAllClassifier> current = make_shared<OneVsAllClassifier>(model);
+
+    ML::Mutable_Feature_Info labelInfo = ML::Mutable_Feature_Info(ML::BOOLEAN);
+
+    int labelValue = 0; 
+
+    std::shared_ptr<Mutable_Feature_Space> fs2 = std::dynamic_pointer_cast<Mutable_Feature_Space>(make_sp(feature_space->make_copy()));
```

Comment:
This won't be true in most cases.  If it's not a mutable feature space, we will need to copy into a new one.